tools/tpm2_createek.c

```diff
@@ -52,59 +108,81 @@ struct createek_context {
-        input_public->publicArea.unique.ecc.y.size = 32;
-        break;
-    case TPM2_ALG_SYMCIPHER:
-        input_public->publicArea.parameters.symDetail.sym.algorithm = TPM2_ALG_AES;
-        input_public->publicArea.parameters.symDetail.sym.keyBits.aes = 128;
-        input_public->publicArea.parameters.symDetail.sym.mode.sym = TPM2_ALG_CFB;
-        input_public->publicArea.unique.sym.size = 0;
-        break;
-    default:
-        LOG_ERR("The algorithm type input(%4.4x) is not supported!",
-                input_public->publicArea.type);
-        return false;
+static const alg_map alg_maps[] = {
+    { "rsa",           "rsa2048:aes128cfb", "sha256", &policy_a_sha256, ATTRS_A },
+    { "rsa2048",       "rsa2048:aes128cfb", "sha256", &policy_a_sha256, ATTRS_A },
+    { "rsa3072",       "rsa3072:aes256cfb", "sha384", &policy_b_sha384, ATTRS_B },
+    { "rsa4096",       "rsa4096:aes256cfb", "sha384", &policy_b_sha384, ATTRS_B },
+    { "ecc",           "ecc_nist_p256:aes128cfb", "sha256", &policy_a_sha256, ATTRS_A },
+    { "ecc256",        "ecc_nist_p256:aes128cfb", "sha256", &policy_a_sha256, ATTRS_A },
+    { "ecc384",        "ecc_nist_p384:aes256cfb", "sha384", &policy_b_sha384, ATTRS_B },
+    { "ecc521",        "ecc_nist_p521:aes256cfb", "sha512", &policy_b_sha512, ATTRS_B },
+    { "ecc_nist_p256", "ecc_nist_p256:aes128cfb", "sha256", &policy_a_sha256, ATTRS_A },
+    { "ecc_nist_p384", "ecc_nist_p384:aes256cfb", "sha384", &policy_b_sha384, ATTRS_B },
+    { "ecc_nist_p521", "ecc_nist_p521:aes256cfb", "sha512", &policy_b_sha512, ATTRS_B },
+    { "ecc_sm2",       "ecc_sm2_p256:sm4_128cfb", "sm3_256", &policy_b_sm3_256, ATTRS_B },
+    { "ecc_sm2_p256",  "ecc_sm2_p256:sm4_128cfb", "sm3_256", &policy_b_sm3_256, ATTRS_B },
+    { "keyedhash",     "xor", "sha256", &policy_a_sha256, ATTRS_A },
+};
+
+static const alg_map *lookup_alg_map(const char *alg) {
+    size_t i;
+
+    for (i = 0; i < ARRAY_LEN(alg_maps); i++)
+    {
+        if (!strcmp(alg, alg_maps[i].input)) {
+            return &alg_maps[i];
+        }
     }
+    return NULL;
+}
 
-    return true;
+static tool_rc init_ek_public(const char *key_alg, TPM2B_PUBLIC *public) {
+
+    const alg_map *m = lookup_alg_map(key_alg);
+
+    if (!m) {
+        LOG_ERR("Invalid key algorithm, got \"%s\"", key_alg);
+        return tool_rc_unsupported;
+    }
+
+    tool_rc rc = tpm2_alg_util_public_init(m->alg, m->namealg, NULL, NULL,
+                                           m->attrs, public);
+    if (rc != tool_rc_success) {
+        return rc;
+    }
+
+    public->publicArea.authPolicy = *m->policy;
+
+    if (public->publicArea.type == TPM2_ALG_ECC &&
+        public->publicArea.parameters.eccDetail.curveID == TPM2_ECC_NIST_P256) {
+        public->publicArea.unique.ecc.x.size = 32;
+        public->publicArea.unique.ecc.y.size = 32;
+    } else  if (public->publicArea.type == TPM2_ALG_RSA &&
+                public->publicArea.parameters.rsaDetail.keyBits == 2048) {
```

Comment:
why only do we care on 2048 RSA and ECCP256, can't their be other keys?

Reply:
This is a requirement in the `TCG EK Credential Profile` for `EK Templates in the Low Range` (B.3.3 Template L-1: RSA 2048) and B.3.4 Template L-2: ECC NIST P256). For the newer `EK Templates in the High Range` the `unique size` must be initialized to `0`.

Reply:
Ahh yeah, I remember that fun, I'm seeing pop back up over in the Python code.